modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableListWrapper.java

```diff
@@ -169,40 +169,36 @@ public void remove(int fromIndex, int toIndex) {
 
     @Override
     public boolean removeAll(Collection<?> c) {
+        if (this.isEmpty() || c.isEmpty()) {
+            return false;
+        }
         beginChange();
-        BitSet bs = new BitSet(c.size());
-        for (int i = 0; i < size(); ++i) {
+        boolean removed = false;
+        for (int i = size()-1; i>=0; i--) {
             if (c.contains(get(i))) {
-                bs.set(i);
-            }
-        }
-        if (!bs.isEmpty()) {
-            int cur = size();
-            while ((cur = bs.previousSetBit(cur - 1)) >= 0) {
-                remove(cur);
+                remove(i);
+                removed = true;
             }
         }
         endChange();
-        return !bs.isEmpty();
+        return removed;
     }
 
     @Override
     public boolean retainAll(Collection<?> c) {
+        if (this.isEmpty() || c.isEmpty()) {
+            return false;
+        }
```

Comment:
logic seems wrong: if c is empty this must be cleared

The error shows up in a failing test (but only when running all base.collections tests in Eclipse, not with gradle on the commandline)

the test failure is in ObservableListWithExtractor.testUpdate_retainAll() which - besides exposing the error above - tells us two issues with the tests

a) the normal (that is no extractor) test is incomplete (in not testing retainAll with empty array/collection)
b) the test is not run with gradle because it doesn't adhere to naming conventions ;)

Reply:
I realized I was wrong. Thank you. I would probably need to add a test for this case.

Reply:
I found that I needed to add at least a retainAll test, but I also found that there was no test class to directly test ObservableListWrapper.

Should I add a test for this non-public API class?
When I add it, I'm not going to write test code for methods other than removeAll and retainAll.